cpp/include/cuspatial/experimental/haversine.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/constants.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cuspatial {
+
+/**
+ * brief Compute haversine distances between points in set A to the corresponding points in set B.
+ *
+ * Computes N haversine distances, where N is `std::distance(a_lon_first, a_lon_last)`. One distance
+ * is computed for each (a_lon[i], a_lat[i]) and (b_lon[i], b_lat[i]) point pair. `distance_first`
+ * must be an iterator to output storage allocated for N distances.
+ *
+ * All iterators must have the same floating-point `value_type`.
+ *
+ * https://en.wikipedia.org/wiki/Haversine_formula
+ *
+ * @param[in]  a_lon_first: beginning of range of longitude of points in set A
+ * @param[in]  a_lon_last: end of the range of longitude of points in set A
+ * @param[in]  b_lon_first: beginning of range of longitude of points in set B
+ * @param[in]  b_lat_first: beginning of range of latitude of points in set B
+ * @param[out] distance_first: beginning of output range of haversine distances
+ * @param[in] radius: radius of the sphere on which the points reside. default: 6371.0
+ *            (approximate radius of Earth in km)
+ *
+ * @tparam LonItA must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LatItA must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LonItB must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LatItB must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam OutputIt must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ *
+ * @return Output iterator to the element past the last distance computed.
```

Comment:
I'm a bit confused by this. I thought for N input pairs there would always be N output distances. Is that not the case?

Reply:
I'm modeling it after std::transform.  This is what it does.

Reply:
Huh, TIL that `std::transform` returns anything. I've honestly never noticed that before.